src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,511 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
+                        this.Otherwise = otherwise;
+                    }
+                    public override string ToString() => "[" + string.Join(",", Cases.Select(c => c.value)) + "]";
+                }
+
+                /// <summary>
+                /// A node representing a final destination that requires no further dispatch.
+                /// </summary>
+                internal sealed class LeafDispatchNode : ValueDispatchNode
+                {
+                    public readonly LabelSymbol Label;
+                    public LeafDispatchNode(SyntaxNode syntax, LabelSymbol Label) : base(syntax) => this.Label = Label;
+                    public override string ToString() => "Leaf";
+                }
+
+                /// <summary>
+                /// A node representing a dispatch based on a relational test of the input value by some constant.
+                /// Nodes of this kind are required to be height-balanced when constructed, so that when the full
+                /// decision tree is produced it generates a balanced tree of comparisons.
+                /// </summary>
+                internal sealed class RelationalDispatch : ValueDispatchNode
+                {
+                    private int _height;
+                    protected override int Height => _height;
+                    public readonly ConstantValue Value;
+                    public readonly BinaryOperatorKind Operator;
+                    /// <summary>The side of the test handling lower values. The true side for &lt; and &lt;=, the false side for > and >=.</summary>
+                    private ValueDispatchNode Left { get; set; }
+                    /// <summary>The side of the test handling higher values. The false side for &lt; and &lt;=, the true side for > and >=.</summary>
+                    private ValueDispatchNode Right { get; set; }
+                    private RelationalDispatch(SyntaxNode syntax, ConstantValue value, BinaryOperatorKind op, ValueDispatchNode left, ValueDispatchNode right) : base(syntax)
+                    {
+                        Debug.Assert(op.OperandTypes() != 0);
+                        this.Value = value;
+                        this.Operator = op;
+                        WithLeftAndRight(left, right);
+                    }
+                    public ValueDispatchNode WhenTrue => IsReversed(Operator) ? Right : Left;
+                    public ValueDispatchNode WhenFalse => IsReversed(Operator) ? Left : Right;
+                    public override string ToString() => $"RelationalDispatch.{Height}({Left} {Operator.Operator()} {Value} {Right})";
+                    private static bool IsReversed(BinaryOperatorKind op) => op.Operator() switch { BinaryOperatorKind.GreaterThan => true, BinaryOperatorKind.GreaterThanOrEqual => true, _ => false };
+
+                    private RelationalDispatch WithLeftAndRight(ValueDispatchNode left, ValueDispatchNode right)
```

Comment:
nit: Why call this method `With...` or return an instance, since it is destructive? Consider `void Update(...)`

Reply:
Our convention is that `Update` is nondestructive too.  This would work either destructively or nondestructively.  It was originally nondestructive but then I noticed that the old node was always garbage after this was called (and that property is local to this class), so I made it destructive.

---
In reply to: [400510417](https://github.com/dotnet/roslyn/pull/42313#discussion_r400510417) [](ancestors = 400510417)